### PR TITLE
planner: eliminate if null on non null column

### PR DESCRIPTION
**Pull request by zhexuany**
### What problem does this PR solve? 
This PR resolves https://github.com/pingcap/tidb/issues/7918. 

### What is changed and how it works?
When buildProjection(), we filter `ast.SelectField` using certain rules. For these fields which does not pass the rules entering original logic. 

details will be provided later. 

### Check List 

Tests 

the test case will be added later.

Code changes

Side effects
 - Possible performance regression

Related changes

### CODE REVIEW

**Comment by winoros** [comment]
This can only partially solve it.
`ifnull` is not only generated during plan building, but also generated by some optimization rule.

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -580,28 +580,55 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
...
-	schema := expression.NewSchema(make([]*expression.Column, 0, len(fields))...)
-	oldLen := 0
-	for _, field := range fields {
-		newExpr, np, err := b.rewrite(field.Expr, p, mapper, true)
-		if err != nil {
-			return nil, 0, errors.Trace(err)
+	var remainingFields []*ast.SelectField
+	fieldSet := make([]bool, len(fields))
+	for idx, field := range fields {
+		if strings.Contains(field.Text(), ast.Ifnull) {
```
Comment (XuHuaiyu) [inline planner/core/logical_plan_builder.go]: Can we use `if !bool` to reduce the layers?

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -580,28 +580,55 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
...
+	}
 
-		col := b.buildProjectionField(proj.id, schema.Len()+1, field, newExpr)
-		schema.Append(col)
+	if len(remainingFields) > 0 {
+		proj := LogicalProjection{Exprs: make([]expression.Expression, 0, len(remainingFields))}.init(b.ctx)
+		schema := expression.NewSchema(make([]*expression.Column, 0, len(remainingFields))...)
+		oldLen := 0
+		for _, field := range remainingFields {
+			newExpr, np, err := b.rewrite(field.Expr, p, mapper, true)
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: Do the optimization here may be a better idea.

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -580,28 +580,55 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
...
+	var remainingFields []*ast.SelectField
+	fieldSet := make([]bool, len(fields))
+	for idx, field := range fields {
+		if strings.Contains(field.Text(), ast.Ifnull) {
+			if fnExpr, ok := field.Expr.(*ast.FuncCallExpr); ok {
+				if cnExpr, ok := fnExpr.Args[0].(*ast.ColumnNameExpr); ok {
+					var colFound *model.ColumnInfo
+					if ds, ok := p.(*DataSource); ok {
+						colFound = model.FindColumnInfo(ds.Columns, cnExpr.Name.Name.L)
+						if mysql.HasNotNullFlag(colFound.Flag) {
```
Comment (lzmhhh123) [inline planner/core/logical_plan_builder.go]: Can we extract these judgement to a function? So that it has strong scalability.

**Comment by zhexuany** [comment]
@winoros @lzmhhh123 PTAL

**Comment by lzmhhh123** [comment]
LGTM.

File: planner/core/logical_plan_builder.go [hunk 4]
```diff
@@ -576,6 +576,23 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
 	}
 }
 
+func eliminateIfNullOnNonNullColumn(p LogicalPlan, expr expression.Expression) expression.Expression {
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: We usually use `NotNull` instead of `NonNull`.

File: planner/core/logical_plan_builder.go [hunk 5]
```diff
@@ -585,6 +602,7 @@ func (b *PlanBuilder) buildProjection(p LogicalPlan, fields []*ast.SelectField,
 	oldLen := 0
 	for _, field := range fields {
 		newExpr, np, err := b.rewrite(field.Expr, p, mapper, true)
+		newExpr = eliminateIfNullOnNonNullColumn(p, newExpr)
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: We can add a TODO here: Do this optimization automatically when creating new scalar function.

File: planner/core/logical_plan_builder.go (hunk 4 above)
Comment (zz-jason) [reply planner/core/logical_plan_builder.go]: s/NonNullColumn/NotNullCol/

File: planner/core/logical_plan_builder.go [hunk 6]
```diff
@@ -576,6 +576,23 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
 	}
 }
 
+func eliminateIfNullOnNonNullColumn(p LogicalPlan, expr expression.Expression) expression.Expression {
+	if scalarExpr, ok := expr.(*expression.ScalarFunction); ok {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: The code nesting is too deep, how about:
```go
ds, isDs := p.(*DataSource)
if !isDs {
	return expr
}

scalarExpr, isScalarFunc := expr.(*expression.ScalarFunction);
if !isScalarFunc || !scalarExpr.FuncName.L == ast.Ifnull {
	return expr
}

colRef, isColRef := scalarExpr.GetArgs()[0].(*expression.Column)
if !isColRef {
	return expr
}

colInfo := model.FindColumnInfo(ds.Columns, cExpr.ColName.L)
if !mysql.HasNotNullFlag(colFound.Flag) {
	return expr
}

return colRef
```

Comment (zz-jason) [reply planner/core/logical_plan_builder.go]: another thing, should we recursively check the expression tree to eliminate as much `IfNull` as possible?

Comment (zz-jason) [reply planner/core/logical_plan_builder.go]: for example:

```sql
select a + ifnull(b, 1) from t;
```

File: planner/core/logical_plan_builder.go [hunk 7]
```diff
@@ -576,6 +576,38 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
...
+		}
+
+		colInfo := model.FindColumnInfo(ds.Columns, colRef.ColName.L)
+		if !mysql.HasNotNullFlag(colInfo.Flag) {
+			return expr
+		}
+
+		return colRef
+	}
+	scalarExpr.SetArgs(exprChildren)
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: this can be removed.

File: planner/core/logical_plan_builder.go [hunk 8]
```diff
@@ -576,6 +576,38 @@ func (b *PlanBuilder) buildProjectionField(id, position int, field *ast.SelectFi
...
+	scalarExpr, isScalarFunc := expr.(*expression.ScalarFunction)
+	if !isScalarFunc {
+		return expr
+	}
+	exprChildren := scalarExpr.GetArgs()
+	for i := 0; i < len(exprChildren); i++ {
+		exprChildren[i] = eliminateIfNullOnNotNullCol(p, exprChildren[i])
+	}
+
+	if scalarExpr.FuncName.L == ast.Ifnull {
```
Comment (zz-jason) [inline planner/core/logical_plan_builder.go]: how about:
```go
if scalarExpr.FuncName.L != ast.Ifnull {
	return expr
}
```

**Pull request merged**